### PR TITLE
Support empty shared memory

### PR DESCRIPTION
**Pull request by sagudev**
More complex alternative to #335 

fixes https://github.com/servo/ipc-channel/issues/268

### CODE REVIEW

File: src/platform/macos/mod.rs [hunk 1]
```diff
@@ -871,7 +871,7 @@ impl OsIpcOneShotServer {
 }
 
 pub struct OsIpcSharedMemory {
-    ptr: *mut u8,
+    ptr: Option<NonNull<u8>>,
     length: usize,
```
Comment (jschwe) [inline src/platform/macos/mod.rs:875]: Does shared memory of size 0 make any sense? How about using [NonZeroUsize](https://doc.rust-lang.org/nightly/std/num/type.NonZeroUsize.html) for the length and just `NonNull<u8>` for the ptr?

Comment (sagudev) [reply src/platform/macos/mod.rs:875]: I was also wondering about that, but then we do not allow empty shared memory. One nicer option would be to simply disallow non-zero `OsIpcSharedMemory` but allow it in `IpcSharedMemory`, so we have:
```rust
pub struct IpcSharedMemory {
    os_shared_memory: Option<OsIpcSharedMemory>,
}
```

**Comment by sagudev** [comment]
I created https://github.com/servo/ipc-channel/pull/335 that handles empty slice in `IpcSharedMemory` instead of in `OsIpcSharedMemory`.

**Comment by sagudev** [comment]
Closed in favor of https://github.com/servo/ipc-channel/pull/335

**Pull request closed without merging**